File: bean-registrar/src/main/java/com/example/demo/PostHandler.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.example.demo;

import org.springframework.web.reactive.function.BodyExtractors;
import org.springframework.web.reactive.function.server.ServerRequest;
import org.springframework.web.reactive.function.server.ServerResponse;
import reactor.core.publisher.Mono;

import java.net.URI;
import java.util.UUID;

/**
 * @author hantsy
 */
public class PostHandler {

    private final PostRepository posts;

    public PostHandler(PostRepository posts) {
        this.posts = posts;
    }

    public Mono<ServerResponse> all(ServerRequest req) {
        return ServerResponse.ok().body(this.posts.findAll(), Post.class);
    }

    public Mono<ServerResponse> create(ServerRequest req) {
        return req.body(BodyExtractors.toMono(Post.class))
                .flatMap(this.posts::save)
                .flatMap(id -> ServerResponse.created(URI.create("/posts/" + id)).build());
    }

    public Mono<ServerResponse> get(ServerRequest req) {
        return this.posts.findById(UUID.fromString(req.pathVariable("id")))
                .flatMap(post -> ServerResponse.ok().body(Mono.just(post), Post.class))
                .switchIfEmpty(ServerResponse.notFound().build());
    }

    public Mono<ServerResponse> update(ServerRequest req) {
        UUID id = UUID.fromString(req.pathVariable("id"));
        return this.posts.findById(id)
                .zipWith(req.bodyToMono(Post.class), (existed, post) -> new Post(existed.id(), post.title(), post.content()))
                .flatMap(post -> this.posts.update(id, post))
                .flatMap(p -> ServerResponse.noContent().build())
                .switchIfEmpty(ServerResponse.notFound().build());
    }

    public Mono<ServerResponse> delete(ServerRequest req) {
        return this.posts.deleteById(UUID.fromString(req.pathVariable("id")))
                .flatMap(id -> ServerResponse.noContent().build());
    }
}

File: bean-registrar-kotlin-dsl/src/main/kotlin/com/example/demo/PostHandler.kt
package com.example.demo

import org.springframework.web.reactive.function.server.*
import java.net.URI
import java.util.*


class PostHandler(val posts: PostRepository) {

    suspend fun all(req: ServerRequest): ServerResponse {
        return ServerResponse.ok().bodyAndAwait(this.posts.findAll())
    }

    suspend fun create(req: ServerRequest): ServerResponse {
        val body = req.awaitBody(Post::class)
        val saved = this.posts.save(body)
        return ServerResponse.created(URI.create("/posts/$saved")).buildAndAwait()
    }

    suspend fun get(req: ServerRequest): ServerResponse {
        val id = UUID.fromString(req.pathVariable("id"))
        val found = this.posts.findById(id)
        return found?.let { ServerResponse.ok().bodyValueAndAwait(found) }
            ?: ServerResponse.notFound().buildAndAwait()
    }

    suspend fun update(req: ServerRequest): ServerResponse {
        val body = req.awaitBody(Post::class)
        val id = UUID.fromString(req.pathVariable("id"))
        return if (this.posts.update(id, body) > 0) {
            ServerResponse.noContent().buildAndAwait()
        } else {
            ServerResponse.notFound().buildAndAwait()
        }

    }

    suspend fun delete(req: ServerRequest): ServerResponse {
        val id = UUID.fromString(req.pathVariable("id"))
        return if (this.posts.deleteById(id) > 0) {
            ServerResponse.noContent().buildAndAwait()
        } else {
            ServerResponse.notFound().buildAndAwait()
        }

    }
}